File: py/packages/over-redis/src/clamator_over_redis/server_transport.py
from __future__ import annotations
import asyncio
import json
import uuid
from typing import Any

from redis.asyncio import Redis

from clamator_protocol import (
    parse_envelope, ClamatorTransportError, Dispatcher,
)
from .keys import command_stream, consumer_group_name, consumer_name


class ServerRedisTransport:
    def __init__(
        self, *, redis: Redis, key_prefix: str,
        instance_id: str | None = None,
        consumer_claim_idle_ms: int = 60_000,
        reply_stream_maxlen: int = 1024,
        shutdown_grace_ms: int = 5_000,
    ) -> None:
        self._redis = redis
        self._key_prefix = key_prefix
        self.instance_id = instance_id or str(uuid.uuid4())
        self._claim_idle_ms = consumer_claim_idle_ms
        self._reply_stream_maxlen = reply_stream_maxlen
        self._shutdown_grace_ms = shutdown_grace_ms
        self._dispatchers: dict[str, Dispatcher] = {}
        self._state = "idle"
        self._tasks: list[asyncio.Task[Any]] = []

    async def register_service(self, name: str, dispatch: Dispatcher) -> None:
        self._dispatchers[name] = dispatch

    async def send(self, env: dict[str, Any], *, timeout: float) -> dict[str, Any]:
        raise NotImplementedError("server transport cannot send requests")

    async def notify(self, env: dict[str, Any]) -> None:
        raise NotImplementedError("server transport cannot send notifications")

    async def start(self) -> None:
        if self._state == "stopped":
            raise ClamatorTransportError("transport has been stopped")
        if self._state == "started":
            return
        self._state = "started"
        for service in self._dispatchers:
            stream = command_stream(self._key_prefix, service)
            group = consumer_group_name(service)
            try:
                await self._redis.xgroup_create(stream, group, id="$", mkstream=True)
            except Exception as e:
                if "BUSYGROUP" not in str(e):
                    raise
            self._tasks.append(asyncio.create_task(self._consumer_loop(service)))
            self._tasks.append(asyncio.create_task(self._reclaim_loop(service)))

    async def stop(self) -> None:
        if self._state != "started":
            self._state = "stopped"
            return
        self._state = "stopped"
        for t in self._tasks:
            t.cancel()
        try:
            await asyncio.wait_for(
                asyncio.gather(*self._tasks, return_exceptions=True),
                timeout=self._shutdown_grace_ms / 1000,
            )
        except asyncio.TimeoutError:
            pass
        self._tasks.clear()

    async def _consumer_loop(self, service: str) -> None:
        stream = command_stream(self._key_prefix, service)
        group = consumer_group_name(service)
        consumer = consumer_name(service, self.instance_id)
        while self._state == "started":
            try:
                results = await self._redis.xreadgroup(
                    group, consumer, {stream: ">"}, block=1000, count=16,
                )
                if not results:
                    continue
                for _, entries in results:
                    for entry_id, fields in entries:
                        await self._handle_entry(service, stream, group, entry_id, fields)
            except asyncio.CancelledError:
                return
            except Exception:
                await asyncio.sleep(0.1)

    async def _reclaim_loop(self, service: str) -> None:
        stream = command_stream(self._key_prefix, service)
        group = consumer_group_name(service)
        consumer = consumer_name(service, self.instance_id)
        while self._state == "started":
            try:
                await asyncio.sleep(max(1.0, self._claim_idle_ms / 4000))
                claimed = await self._redis.xautoclaim(
                    stream, group, consumer, min_idle_time=self._claim_idle_ms, count=32,
                )
                # claimed: (next_cursor, [(id, fields), ...], deleted_ids)
                _, entries, _ = claimed if len(claimed) == 3 else (claimed[0], claimed[1], [])
                for entry_id, fields in entries or []:
                    await self._handle_entry(service, stream, group, entry_id, fields)
            except asyncio.CancelledError:
                return
            except Exception:
                pass

    async def _handle_entry(
        self, service: str, stream: str, group: str, entry_id: Any, fields: dict[Any, Any],
    ) -> None:
        env_b = fields.get(b"envelope") or fields.get("envelope")
        if env_b is None:
            await self._redis.xack(stream, group, entry_id)
            return
        if isinstance(env_b, bytes):
            env_b = env_b.decode()
        reply_to = fields.get(b"reply-to") or fields.get("reply-to")
        if isinstance(reply_to, bytes):
            reply_to = reply_to.decode()
        try:
            env_obj = json.loads(env_b)
            parsed = parse_envelope(env_obj)
        except Exception:
            await self._redis.xack(stream, group, entry_id)
            return
        dispatcher = self._dispatchers.get(service)
        if dispatcher is None:
            await self._redis.xack(stream, group, entry_id)
            return
        reply = await dispatcher(parsed)
        if reply_to and reply is not None:
            await self._redis.xadd(
                reply_to, {"type": "rpc", "envelope": json.dumps(reply)},
                maxlen=self._reply_stream_maxlen, approximate=True,
            )
        await self._redis.xack(stream, group, entry_id)

File: ts/packages/over-redis/src/server-transport.ts
import { randomUUID } from 'node:crypto';
import IORedis, { type Redis } from 'ioredis';
import {
  parseEnvelope, EnvelopeKind, ClamatorTransportError,
  type Transport, type Dispatcher, type SendOptions,
} from '@clamator/protocol';
import { commandStream, consumerGroupName, consumerName } from './keys.js';

export interface ServerTransportOptions {
  /** Existing ioredis instance. Mutually exclusive with `redisUrl`. */
  redis?: Redis;
  /** Redis URL. If neither `redis` nor `redisUrl` is provided, falls back to `process.env.REDIS_URL` then `redis://localhost:6379`. */
  redisUrl?: string;
  keyPrefix: string;
  instanceId?: string;
  consumerClaimIdleMs?: number;
  replyStreamMaxLen?: number;
  shutdownGraceMs?: number;
}

export class ServerRedisTransport implements Transport {
  readonly instanceId: string;
  private dispatchers = new Map<string, Dispatcher>();
  private state: 'idle' | 'started' | 'stopped' = 'idle';
  private loops: Promise<void>[] = [];
  private abort = false;
  private reclaimWakers: Set<() => void> = new Set();
  // Dedicated connection per service for the blocking XREADGROUP call so that
  // xadd (reply), xack, and xautoclaim on the main connection are never
  // queued behind the blocking read.
  private blockingConns: Redis[] = [];

  private readonly redis: Redis;
  private readonly ownsRedis: boolean;
  private readonly keyPrefix: string;
  private readonly replyStreamMaxLen: number;
  private readonly consumerClaimIdleMs: number;
  private readonly shutdownGraceMs: number;

  constructor(opts: ServerTransportOptions) {
    if (opts.redis && opts.redisUrl)
      throw new ClamatorTransportError('provide either `redis` or `redisUrl`, not both');
    if (opts.redis) {
      this.redis = opts.redis;
      this.ownsRedis = false;
    } else {
      const url = opts.redisUrl ?? process.env.REDIS_URL ?? 'redis://localhost:6379';
      this.redis = new IORedis(url);
      this.ownsRedis = true;
    }
    this.keyPrefix = opts.keyPrefix;
    this.instanceId = opts.instanceId ?? randomUUID();
    this.replyStreamMaxLen = opts.replyStreamMaxLen ?? 1024;
    this.consumerClaimIdleMs = opts.consumerClaimIdleMs ?? 60_000;
    this.shutdownGraceMs = opts.shutdownGraceMs ?? 5_000;
  }

  async registerService(name: string, dispatch: Dispatcher): Promise<void> {
    this.dispatchers.set(name, dispatch);
  }

  async send(): Promise<Record<string, unknown>> {
    throw new Error('server transport cannot send requests');
  }

  async notify(): Promise<void> {
    throw new Error('server transport cannot send notifications');
  }

  async start(): Promise<void> {
    if (this.state === 'stopped') throw new ClamatorTransportError('transport has been stopped');
    if (this.state === 'started') return;
    this.state = 'started';
    this.abort = false;
    for (const service of this.dispatchers.keys()) {
      const stream = commandStream(this.keyPrefix, service);
      const group = consumerGroupName(service);
      try {
        await this.redis.xgroup('CREATE', stream, group, '$', 'MKSTREAM');
      } catch (e) {
        const msg = (e as Error).message;
        if (!msg.includes('BUSYGROUP')) throw e;
      }
      // Dedicated connection for blocking XREADGROUP so the main connection
      // stays free for non-blocking operations (xadd, xack, xautoclaim).
      const blockingRedis = this.redis.duplicate();
      this.blockingConns.push(blockingRedis);
      this.loops.push(this.runConsumerLoop(service, blockingRedis));
      this.loops.push(this.runReclaimLoop(service));
    }
  }

  async stop(): Promise<void> {
    if (this.state !== 'started') { this.state = 'stopped'; return; }
    this.abort = true;
    // Wake up any sleeping reclaim loops so they can check abort and exit promptly.
    for (const wake of this.reclaimWakers) wake();
    this.reclaimWakers.clear();
    // Disconnect blocking connections to unblock XREADGROUP immediately.
    for (const conn of this.blockingConns) {
      try { conn.disconnect(); } catch { /* ignore */ }
    }
    const grace = this.shutdownGraceMs;
    await Promise.race([
      Promise.allSettled(this.loops),
      new Promise(r => setTimeout(r, grace)),
    ]);
    // Clean up dedicated blocking connections.
    for (const conn of this.blockingConns) {
      try { await conn.quit(); } catch { /* best effort */ }
    }
    this.blockingConns = [];
    if (this.ownsRedis) {
      try { await this.redis.quit(); } catch { /* best effort */ }
    }
    this.state = 'stopped';
  }

  private async runConsumerLoop(service: string, blockingRedis: Redis): Promise<void> {
    const stream = commandStream(this.keyPrefix, service);
    const group = consumerGroupName(service);
    const consumer = consumerName(service, this.instanceId);
    while (!this.abort) {
      try {
        const result = await blockingRedis.xreadgroup(
          'GROUP', group, consumer, 'COUNT', 16, 'BLOCK', 1000,
          'STREAMS', stream, '>',
        );
        if (!result) continue;
        for (const [, entries] of result as [string, [string, string[]][]][]) {
          for (const [entryId, fields] of entries) {
            await this.handleEntry(service, stream, group, entryId, fields);
          }
        }
      } catch (err) {
        if (this.abort) return;
        await new Promise(r => setTimeout(r, 100));
      }
    }
  }

  private interruptibleSleep(ms: number): Promise<void> {
    return new Promise<void>(resolve => {
      const timer = setTimeout(() => { this.reclaimWakers.delete(wake); resolve(); }, ms);
      const wake = () => { clearTimeout(timer); this.reclaimWakers.delete(wake); resolve(); };
      this.reclaimWakers.add(wake);
    });
  }

  private async runReclaimLoop(service: string): Promise<void> {
    const stream = commandStream(this.keyPrefix, service);
    const group = consumerGroupName(service);
    const consumer = consumerName(service, this.instanceId);
    const idleThreshold = this.consumerClaimIdleMs;
    while (!this.abort) {
      try {
        await this.interruptibleSleep(Math.max(1000, idleThreshold / 4));
        if (this.abort) return;
        const claimed = await this.redis.xautoclaim(
          stream, group, consumer, idleThreshold, '0', 'COUNT', 32,
        ) as [string, [string, string[]][], string[]];
        const entries = claimed[1];
        if (!entries || entries.length === 0) continue;
        for (const [entryId, fields] of entries) {
          await this.handleEntry(service, stream, group, entryId, fields);
        }
      } catch (err) {
        if (this.abort) return;
      }
    }
  }

  private async handleEntry(
    service: string, stream: string, group: string,
    entryId: string, fields: string[],
  ): Promise<void> {
    const envIdx = fields.indexOf('envelope');
    if (envIdx < 0) {
      await this.redis.xack(stream, group, entryId);
      return;
    }
    const replyToIdx = fields.indexOf('reply-to');
    const replyTo = replyToIdx >= 0 ? fields[replyToIdx + 1] : null;
    let envObj: Record<string, unknown>;
    try { envObj = JSON.parse(fields[envIdx + 1] ?? '') as Record<string, unknown>; }
    catch { await this.redis.xack(stream, group, entryId); return; }
    let parsed;
    try { parsed = parseEnvelope(envObj); } catch { await this.redis.xack(stream, group, entryId); return; }

    const dispatcher = this.dispatchers.get(service);
    if (!dispatcher) { await this.redis.xack(stream, group, entryId); return; }

    const reply = await dispatcher(parsed);
    if (replyTo && reply) {
      await this.redis.xadd(
        replyTo, 'MAXLEN', '~', String(this.replyStreamMaxLen), '*',
        'type', 'rpc', 'envelope', JSON.stringify(reply),
      );
    }
    await this.redis.xack(stream, group, entryId);
  }
}
